lib/src/tokenizer.cc

```diff
@@ -0,0 +1,556 @@
+#include <internal/tokenizer.hpp>
+#include <internal/config_util.hpp>
+#include <internal/values/config_boolean.hpp>
+#include <internal/values/config_null.hpp>
+#include <internal/values/config_double.hpp>
+#include <internal/values/config_long.hpp>
+#include <internal/values/config_string.hpp>
+#include <boost/nowide/fstream.hpp>
+#include <boost/lexical_cast.hpp>
+
+using namespace std;
+
+namespace hocon {
+
+    problem_exception::problem_exception(problem prob) :
+            runtime_error(prob.message()), _problem(move(prob)) { }
+
+    problem const& problem_exception::get_problem() const {
+        return _problem;
+    }
+
+    /** Whitespace Saver */
+    token_iterator::whitespace_saver::whitespace_saver() : _last_token_was_simple_value(false) { }
+
+    void token_iterator::whitespace_saver::add(char c) {
+        _whitespace += c;
+    }
+
+    shared_ptr<token> token_iterator::whitespace_saver::check(token_type type,
+                                                              simple_config_origin base_origin,
+                                                              int line_number)
+    {
+        if (is_simple_value(type)) {
+            return next_is_simple_value(base_origin, line_number);
+        } else {
+            return next_is_not_simple_value(base_origin, line_number);
+        }
+    }
+
+    /**
+     * Called if the next token is not a simple value;
+     * discards any whitespace we were saving between
+     * simple values.
+     */
+    shared_ptr<token> token_iterator::whitespace_saver::next_is_not_simple_value(
+            simple_config_origin base_origin, int line_number)
```

Comment:
This could probably take by const&, since it's just going to use it to create a copy (not take ownership of it).